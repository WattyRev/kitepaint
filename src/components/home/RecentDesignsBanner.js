import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import designShape from "../../models/design";
import { H2, Button, P } from "../../theme";

const StyleWrapper = styled.div`
  padding: 16px;
  > .designs {
    display: flex;
    justify-content: space-evenly;
    flex-wrap: wrap;
  }
  .design-preview {
    width:200px
    height: 200px;
    flex-shrink: 0;
    padding: 16px 0;

    > svg {
      position: relative;
      top: 50%;
      transform: translateY(-50%) rotate(45deg);
    }
  }
  > .see-all-wrapper {
    max-width: 400px;
    margin: 0 auto;
  }
  .design-name {
    position: absolute;
    top: 10%;
    left: 0;
    padding: 4px;
    background: rgba(0,0,0,.8);
    transform: translateX(-110%);
    transition: .5s transform;
  }
  .design-wrapper {
    position: relative;
    overflow: hidden;
    cursor: pointer;

    &:hover {
      .design-name {
        transform: translateX(0);
      }
    }
  }
`;

/**
 * The banner displaying the recently created public designs.
 */
const RecentDesignsBanner = ({ designs }) => (
  <StyleWrapper>
    <H2>Recent Designs</H2>
    <div className="designs">
      {designs.map(design => {
        return (
          <Link
            key={design.id}
            className="design-wrapper"
            to={`/view/${design.id}`}
          >
            <div
              className="design-preview"
              dangerouslySetInnerHTML={{
                __html: design.variations.find(variation => variation.primary)
                  .svg
              }}
            />
            <P isLight className="design-name">
              {design.name}
            </P>
          </Link>
        );
      })}
    </div>
    <div className="see-all-wrapper">
      <Button isBlock as={Link} to="/designs">
        See All Designs
      </Button>
    </div>
  </StyleWrapper>
);

RecentDesignsBanner.propTypes = {
  /**
   * The designs to be displayed
   */
  designs: PropTypes.arrayOf(designShape).isRequired
};

export default RecentDesignsBanner;
