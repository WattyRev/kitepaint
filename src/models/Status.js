import PropTypes from "prop-types";

/**
 * Status can be applied to any model to indicate it's public visibility.
 */
export const statusProp = PropTypes.oneOf(["0", "1", "2"]);
const Status = {
  "0": "Private",
  "1": "Unlisted",
  "2": "Public",
  PRIVATE: "0",
  UNLISTED: "1",
  PUBLIC: "2"
};
export default Status;
