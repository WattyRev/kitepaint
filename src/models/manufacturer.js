import PropTypes from "prop-types";

/**
 * A manufacturer is a company that manufactures kites and owns one or more products.
 */
const manufacturerState = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  logo: PropTypes.string.isRequired,
  website: PropTypes.string
});

export default manufacturerState;

/**
 * A mock manufacturer used for testing
 * @type {Object}
 */
const getMockManufacturer = () => ({
  id: "abc",
  name: "Krazy Kites",
  logo: "abc.jpg",
  website: "http://krazykites.com"
});
export { getMockManufacturer };
