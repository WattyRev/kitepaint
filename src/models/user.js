import PropTypes from "prop-types";

/**
 * A user is an individual user account. It contains limited details that are available publically.
 */
const userShape = PropTypes.shape({
  username: PropTypes.string.isRequired,
  loginid: PropTypes.string.isRequired
});

export default userShape;

/**
 * A mock user used for testing.
 */
const getMockUser = () => ({
  username: "FlyFucker",
  loginid: "abc"
});

export { getMockUser };
