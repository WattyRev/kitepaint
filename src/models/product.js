import PropTypes from "prop-types";
import { getMockManufacturer } from "./manufacturer";
import Status, { statusProp } from "./status";

/**
 * A product is a kite that can be customized to create a design.
 */
const productShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  embed: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  notes: PropTypes.arrayOf(PropTypes.string),
  manufacturer: PropTypes.string.isRequired,
  status: statusProp.isRequired,
  url: PropTypes.string,
  colors: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired
    })
  ),
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired
    })
  )
});

export default productShape;

/**
 * A mock product used for testing
 * @type {Object}
 */
const getMockProduct = () => ({
  id: "product-1",
  embed: "krazykites.com",
  name: "Krazy Kite",
  manufacturer: getMockManufacturer().id,
  url: "http://krazykites.com/krazy-kite",
  colors: [
    {
      name: "red",
      color: "#ff0000"
    },
    {
      name: "black",
      color: "#000000"
    }
  ],
  variations: [
    {
      name: "Standard",
      svg: "<div>Kool Kite</div>"
    }
  ],
  status: Status.PUBLIC
});

export { getMockProduct };
