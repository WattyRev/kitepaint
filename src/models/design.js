import PropTypes from "prop-types";
import { getMockProduct } from "./product";
import Status, { statusProp } from "./status";

/**
 * A design is a user-created entity. It represetns a saved design of a particular product.
 */
const designShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  // A string representing the created date
  created: PropTypes.string.isRequired,
  // A string representing the updated date
  updated: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  // The ID of the cooresponding product
  product: PropTypes.string.isRequired,
  // The ID of the user that created the design
  user: PropTypes.string,
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired,
      primary: PropTypes.bool.isRequired
    })
  ),
  // The status of the cooresponding product
  productStatus: statusProp.isRequired,
  // The status of the design
  status: statusProp.isRequired
});

export default designShape;

/**
 * A mock design used for testing
 * @type {Object}
 */
const getMockDesign = () => ({
  id: "design-1",
  created: "01/01/2018",
  updated: "01/02/2018",
  name: "picasso",
  product: getMockProduct().id,
  variations: [
    {
      name: "Standard",
      svg: "<div>picasso standard</div>",
      primary: true
    }
  ],
  status: Status.PRIVATE,
  productStatus: Status.PUBLIC
});
export { getMockDesign };
