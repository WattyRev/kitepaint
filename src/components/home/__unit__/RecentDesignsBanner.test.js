import React from "react";
import { shallow } from "enzyme";
import { getMockDesign } from "../../../models/design";
import RecentDesignsBanner from "../RecentDesignsBanner";

describe("RecentDesignsBanner", () => {
  let props;
  beforeEach(() => {
    props = {
      designs: [getMockDesign()]
    };
  });
  it("renders", () => {
    expect.assertions(1);
    const wrapper = shallow(<RecentDesignsBanner {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
