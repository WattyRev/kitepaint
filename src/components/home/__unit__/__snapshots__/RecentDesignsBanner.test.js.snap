// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentDesignsBanner renders 1`] = `
<styled.div>
  <Styled(H2)>
    Recent Designs
  </Styled(H2)>
  <div
    className="designs"
  >
    <Link
      className="design-wrapper"
      key="123"
      replace={false}
      to="/view/123"
    >
      <div
        className="design-preview"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<div class=\\"totally-svg\\">Pretty Design</div>",
          }
        }
      />
      <P
        className="design-name"
        isLight={true}
      >
        abc
      </P>
    </Link>
  </div>
  <div
    className="see-all-wrapper"
  >
    <Button
      as={[Function]}
      isBlock={true}
      to="/designs"
    >
      See All Designs
    </Button>
  </div>
</styled.div>
`;
